### PR TITLE
Fix RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE in try-with-resources in presence of debug symbols

### LINKED ISSUE #1931: debug symbols trigger false positive RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE on java11
Starting 4.4.0 and up to at least 4.5.3 there's a regression described below.
Especially annoying as is block log4j CVE mitigation upgrade to 4.5.3

```java
import java.util.stream.Stream;

public class A {
  private int foo() {
    try (final Stream<Object> stream = Stream.of()) {
      return stream.hashCode();
    }
  }
}
```
with debug symbols
```console
$ javac --version
javac 11.0.9
$ javac -g A.java && jar -cf A.jar A.class
$ java -cp spotbugs.jar edu.umd.cs.findbugs.LaunchAppropriateUI -textui -longBugCodes A.jar
M C RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE RCN: Nullcheck of stream at line 6 of value previously dereferenced in A.foo()  At A.java:[line 6]
```
without debug symbols
```console
$ javac A.java && jar -cf A.jar A.class
$ java -cp spotbugs.jar edu.umd.cs.findbugs.LaunchAppropriateUI -textui -longBugCodes A.jar
```

Output of `javap -l -p -c -v` related to method `foo`
```
  private int foo();
    descriptor: ()I
    flags: (0x0002) ACC_PRIVATE
    Code:
      stack=2, locals=4, args_size=1
         0: iconst_0
         1: anewarray     #2                  // class java/lang/Object
         4: invokestatic  #3                  // InterfaceMethod java/util/stream/Stream.of:([Ljava/lang/Object;)Ljava/util/stream/Stream;
         7: astore_1
         8: aload_1
         9: invokevirtual #4                  // Method java/lang/Object.hashCode:()I
        12: istore_2
        13: aload_1
        14: ifnull        23
        17: aload_1
        18: invokeinterface #5,  1            // InterfaceMethod java/util/stream/Stream.close:()V
        23: iload_2
        24: ireturn
        25: astore_2
        26: aload_1
        27: ifnull        45
        30: aload_1
        31: invokeinterface #5,  1            // InterfaceMethod java/util/stream/Stream.close:()V
        36: goto          45
        39: astore_3
        40: aload_2
        41: aload_3
        42: invokevirtual #7                  // Method java/lang/Throwable.addSuppressed:(Ljava/lang/Throwable;)V
        45: aload_2
        46: athrow
      Exception table:
         from    to  target type
             8    13    25   Class java/lang/Throwable
            30    36    39   Class java/lang/Throwable
      LineNumberTable:
        line 5: 0
        line 6: 8
        line 7: 13
        line 6: 23
        line 5: 25
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            8      39     1 stream   Ljava/util/stream/Stream;
            0      47     0  this   LA;
      LocalVariableTypeTable:
        Start  Length  Slot  Name   Signature
            8      39     1 stream   Ljava/util/stream/Stream<Ljava/lang/Object;>;
      StackMapTable: number_of_entries = 4
        frame_type = 253 /* append */
          offset_delta = 23
          locals = [ class java/util/stream/Stream, int ]
        frame_type = 255 /* full_frame */
          offset_delta = 1
          locals = [ class A, class java/util/stream/Stream ]
          stack = [ class java/lang/Throwable ]
        frame_type = 255 /* full_frame */
          offset_delta = 13
          locals = [ class A, class java/util/stream/Stream, class java/lang/Throwable ]
          stack = [ class java/lang/Throwable ]
        frame_type = 5 /* same */
```
the difference is that debug version has this block
```
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            8      39     1 stream   Ljava/util/stream/Stream;
            0      47     0  this   LA;
      LocalVariableTypeTable:
        Start  Length  Slot  Name   Signature
            8      39     1 stream   Ljava/util/stream/Stream<Ljava/lang/Object;>;
```
and triggers false positive, while non-debug version doesn't have this block and works

related to
https://github.com/spotbugs/spotbugs/issues/868
https://github.com/spotbugs/spotbugs/issues/1338
https://github.com/spotbugs/spotbugs/issues/1694

likely introduced in 07d001ae705f9ec018331f5d88c824b7a372626f #1575

### PR DESCRIPTION
**Pull request by dmivankov**
Resolves #1931 by improving on #1575
Probably also #868, #1338, #1694

Not entirely sure about what exactly is happening but since 4.4.0 on javac11+ (openjdk)
findbugs was reporting redundant null check if all of following holds
- code in try block always dereferences the object
- try-with-resources is applied to an object via interface reference
- debug symbols are present (LocalVariableTable, via javac -g)

javac11+ generates bytecode along the lines of
```
try-body
if (obj != null) obj.close()
on_any_exceptions_in_try_body: if (obj != null) obj.close()
```
This does always close obj if it is not null
Javac could've deduced that if try-body always throws on `obj==null` then
null check isn't needed but isn't obliged to do so, for simplicity for example.

Fix by treating invokeinterface same way as invokevirtual in FindNullDeref
Adding a minimal test for interface case.

Note that test suite is compiled with default gradle settings that have debug
symbols enabled.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Review by ThrawnCA — APPROVED** [review]

**Review by ThrawnCA — DISMISSED** [review]

**Comment by dmivankov** [comment]
I don't have write access to click merge, would appreciate to have it eventually merged and released :)

**Comment by dmivankov** [comment]
@ThrawnCA can this be released in 4.6.1+? Fixing false positive regression would make it easier to bump spotbugs to versions with log4j security issues fixed

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java [hunk 1]
```diff
@@ -1919,6 +1920,21 @@ private boolean isGeneratedCodeInCatchBlockViaLineNumber(@NonNull ConstantPool c
...
+                        if (throwables.stream().anyMatch(x -> lineNumberTable.getSourceLine(x.getEndPC()) == currentLine)) {
+                            closeCounter++;
+                        }
+                    } else if ("addSuppressed".equals(methodName)) {
+                        if (relevantLineNumbers.contains(currentLine)) {
+                            addSuppressedPresent = true;
+                        }
+                    }
+                }
+                break;
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: Even though I remove these lines, the `./gradlew build` was completed without test failure. Does it mean that the test case you added is not enough to reproduce the issue?

**Review by KengoTODA — CHANGES_REQUESTED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java (hunk 1 above)
Comment (dmivankov) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: Let me check ,it was important whether test is compiled with debug symbols or without them

Comment (KengoTODA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: if possible, configure Gradle's JavaCompile tasks to make the test reproducible.

Comment (dmivankov) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: I think I was able to reproduce it before via spotbugs tests, but maybe something changed there.
For now can reproduce in following environment:
```console
$ gradle --version

------------------------------------------------------------
Gradle 7.3
------------------------------------------------------------

Build time:   2021-11-09 20:40:36 UTC
Revision:     96754b8c44399658178a768ac764d727c2addb37

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          17.0.1 (Oracle Corporation 17.0.1+12-nixos)
OS:           Linux 5.10.81 amd64
```
```console
$ java --version
openjdk 11.0.9 2020-10-20
OpenJDK Runtime Environment (build 11.0.9+0-adhoc..source)
OpenJDK 64-Bit Server VM (build 11.0.9+0-adhoc..source, mixed mode)
```
by running
```console
$ git checkout origin/master spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java
$ gradle assemble
$ java -cp "./spotbugs/build/libs/spotbugs.jar:./spotbugs-annotations/build/libs/*:./spotbugs/.libs/*"  edu.umd.cs.findbugs.LaunchAppropriateUI -textui -longBugCodes -onlyAnalyze "ghIssues.issue600.*" spotbugsTestCases/build/libs/spotbugsTestCases-*.jar
M C RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE RCN: Nullcheck of x at line 7 of value previously dereferenced in ghIssues.issue600.TryWithResourcesSimple.nocatchInterface(IMyAutocloseable)  At TryWithResourcesSimple.java:[line 6]
...

can try to debug why `gradle test` doesn't fail on new test without the fix, but going to take a bit of time

Comment (dmivankov) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: invocation for just issue600 classes
> java -cp "./spotbugs/build/libs/spotbugs.jar:./spotbugs-annotations/build/libs/*:./spotbugs/.libs/*"  edu.umd.cs.findbugs.LaunchAppropriateUI -textui -longBugCodes -onlyAnalyze "ghIssues.issue600.*" -auxclasspath spotbugsTestCases/build/classes/java/main/ghIssues/issue600/MyAutocloseable.class:spotbugsTestCases/build/classes/java/main/ghIssues/issue600/IMyAutocloseable.class  spotbugsTestCases/build/classes/java/main/ghIssues/issue600/TryWithResourcesSimple.class 
>
> M C RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE RCN: Nullcheck of x at line 7 of value previously dereferenced in ghIssues.issue600.TryWithResourcesSimple.nocatchInterface(IMyAutocloseable)  At TryWithResourcesSimple.java:[line 6]

Comment (dmivankov) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: `RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE` before vs `RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE` now, will update test to account for that

Comment (dmivankov) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: Updated the test, now it fails for me without the fix

Comment (KengoTODA) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1937]: I also confirmed that `./gradlew build` fails if I remove these lines.
Thanks!

**Review by KengoTODA — CHANGES_REQUESTED** [review]
Changes are LGTM.
Please run `./gradlew spotlessApply` to format your code.

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/Issue600Test.java [hunk 2]
```diff
@@ -37,14 +37,20 @@ public void testExplicitNulls() {
 
     private void assertBugCount(String clazz, int count) {
         BugCollection bugCollection = analyse(clazz);
+
         final BugInstanceMatcher bugTypeMatcher = new BugInstanceMatcherBuilder()
                 .bugType("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE").build();
         assertThat(bugCollection, containsExactly(count, bugTypeMatcher));
+
+        final BugInstanceMatcher bugTypeMatcher2 = new BugInstanceMatcherBuilder()
+            .bugType("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE").build();
```
Comment (dmivankov) [inline spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/Issue600Test.java:46]:
```suggestion
                .bugType("RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE").build();
```

**Review by KengoTODA — APPROVED** [review]

**Comment by dmivankov** [comment]
@ThrawnCA @KengoTODA anything else I can help with for this PR to be merged and released?

**Comment by dmivankov** [comment]
@KengoTODA I see 4.7.0 released without this fix even though it was approved, how do we get it released?

**Review by ThrawnCA — APPROVED** [review]

**Review by KengoTODA — APPROVED** [review]
LGTM, thank you! The next release (4.7.1 or 4.8.0) will contain this fix.

**Comment by Paharov** [comment]
@dmivankov Hi! I checked if it solves the false positive RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE too for Java 11 yesterday, but unfortunately it didn't, so I think it makes https://github.com/spotbugs/spotbugs/issues/1694 still active.

**Comment by dmivankov** [comment]
Ow, can confirm that also the checked-in test fails in master, investigating

**Pull request merged**